tree_unique_args/src/loop_strength_reduction.egg

```diff
@@ -0,0 +1,111 @@
+(ruleset loop-strength-reduction)
+
+;; Change multiplication of loop variable to addition to an accumulator
+;;
+;; ORIGINAL
+;; a = 0
+;; c = 3
+;; for  i = 0 to n:
+;;     a = i * c
+;;
+;; OPTIMIZED
+;; a = 0
+;; c = 3
+;; d = 0
+;; for i = 0 to n:
+;;     a = d
+;;     d += c
+
+; Finds invariants/constants within a loop.
+; Columns: loop; value of invariant in inputs; value of invariant in outputs
+; TODO: once everything else is working, change this to just use an ID to ref the loop
+(relation Inv (Expr Expr Expr))
+(rule (
+    (= loop (Loop id (All (Parallel) inputs) (All (Sequential) (Pair pred (All (Parallel) outputs)))))
```

Comment:
My understanding is that a `Pair` can be desugared into `Cons/Nil`, but not the other direction, so having `Pair` in a rule's trigger may make the rule incomplete.

Reply:
I think we changed it so that we have the other direction too